util/src/main/java/io/kubernetes/client/util/Cache.java

```diff
@@ -0,0 +1,269 @@
+package io.kubernetes.client.util;
+
+import com.squareup.okhttp.Call;
+import com.squareup.okhttp.ResponseBody;
+import io.kubernetes.client.ApiClient;
+import io.kubernetes.client.ApiException;
+import io.kubernetes.client.models.V1ObjectMeta;
+import java.io.Closeable;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Cache<ApiType, ApiListType> implements Closeable {
+  private Function<CallGeneratorParams, Call> callGenerator;
+  private ApiClient client;
+  private Watch<ApiType> watch;
+  private Type watchType;
+  private ApiListType list;
+  private Class<?> listClass;
+  private Thread watchThread;
+  private Integer timeout;
+  private Map<String, List<ApiType>> namespaceIndex;
+
+  private static final Logger log = LoggerFactory.getLogger(Cache.class);
+
+  private Cache() {}
+
+  public Cache(
+      ApiClient client,
+      Function<CallGeneratorParams, Call> callGenerator,
+      Class<?> listClass,
+      Type watchType) {
+    this.client = client;
+    // TODO: Parameter?
+    this.timeout = 5 * 60;
+    this.callGenerator = callGenerator;
+    this.watch = null;
+    this.watchType = watchType;
+    this.listClass = listClass;
+    this.list = null;
+    this.namespaceIndex = new HashMap<>();
+  }
+
+  private void init() throws ApiException, IOException {
+    if (list == null) {
+      updateList();
+    }
+    if (watch == null) {
+      startWatch();
+    }
+  }
+
+  public synchronized ApiListType list() throws ApiException, IOException {
+    init();
+    return list;
```

Comment:
Consider deep cloning this object since it could be accidentally mutated underneath the user, on the other hand, cloning could be pretty bad for perf....